contracts/tatum/Tatum721Provenance.sol

```diff
@@ -0,0 +1,233 @@
+            if (_cashbackRecipients[tokenId][i] == _msgSender()) {
+                _cashbackValues[tokenId][i] = cashbackValue;
+                return true;
+            }
+        }
+        return true;
+    }
+
+    function burn(uint256 tokenId) public virtual {
+        //solhint-disable-next-line max-line-length
+        require(
+            _isApprovedOrOwner(_msgSender(), tokenId),
+            "ERC721Burnable: caller is not owner nor approved"
+        );
+        _burn(tokenId);
+    }
+    function uintToString(uint v) private pure returns (string memory) {
+        uint maxlength = 100;
+        bytes memory reversed = new bytes(maxlength);
+        uint i = 0;
+        while (v != 0) {
+            uint remainder = v % 10;
+            v = v / 10;
+            reversed[i++] = bytes1(uint8(48 + remainder));
+        }
+        bytes memory s = new bytes(i); // i + 1 is inefficient
+        for (uint j = 0; j < i; j++) {
+            s[j] = reversed[i - j - 1]; // to avoid the off-by-one error
+        }
+        string memory str = string(s);  // memory isn't implicitly convertible to storage
+        return str;
+    }
+    function safeTransfer(
+        address to,
+        uint256 tokenId,
+        string memory data,
+        uint256 value
+    ) public payable {
+        string memory dataValue=uintToString(value);
+        bytes memory bytesData=abi.encodePacked(data,"'''####'''",dataValue);
+        
+        if (_cashbackRecipients[tokenId].length != 0) {
+            // checking cashback addresses exists and sum of cashbacks
+            require(
+                _cashbackRecipients[tokenId].length != 0,
+                "CashbackToken should be of cashback type"
+            );
+            uint256 percentSum = 0;
+            for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+                percentSum += _cashbackValues[tokenId][i];
+            }
+            uint256 sum = (percentSum * value) / 100;
+            if (sum > msg.value) {
+                payable(msg.sender).transfer(msg.value);
+                revert(
+                    "Value should be greater than or equal to cashback value"
+                );
+            }
+            for (uint256 i = 0; i < _cashbackRecipients[tokenId].length; i++) {
+                // transferring cashback to authors
+                if (_cashbackValues[tokenId][i] > 0) {
+                    payable(_cashbackRecipients[tokenId][i]).transfer(
+                        (_cashbackValues[tokenId][i] * value) / 100
+                    );
+                }
+            }
+            if (msg.value > sum) {
+                payable(msg.sender).transfer(msg.value - sum);
+            }
+            _safeTransfer(_msgSender(), to, tokenId, bytesData);
+        } else {
+            _safeTransfer(_msgSender(), to, tokenId, bytesData);
+        }
+        _setTokenData(tokenId,string(bytesData));
+        emit TransferWithProvenance(tokenId, to, string(bytesData));
```

Comment:
you should emit here Transfer and TransferWithProvenance. Transfer event is for backward compatibility.

Reply:
transfer event is emitted by default